### ISSUE TITLE
module-info.java should contain "requires java.logging"

### ISSUE THREAD

**Issue by cowwoc** [classgraph/classgraph#309]
Version 4.6.32

When I run `new ClassGraph().verbose().scan()` under IntellJ IDEA I get:

```
Exception in thread "main" java.lang.IllegalAccessError: class nonapi.io.github.classgraph.utils.LogNode (in module io.github.classgraph) cannot access class java.util.logging.Logger (in module java.logging) because module io.github.classgraph does not read module java.logging
	at io.github.classgraph@4.6.32/nonapi.io.github.classgraph.utils.LogNode.<clinit>(LogNode.java:51)
	at io.github.classgraph@4.6.32/io.github.classgraph.ClassGraph.verbose(ClassGraph.java:103)
```

I believe adding `requires java.logging` to your `module-info.java` file will fix this problem.

**Comment by GedMarc** [comment]
Let me check this please - I go through a lot of JPMS with classgraph, have never seen this

**Comment by GedMarc** [comment]
(y) verbose logging

**Comment by lukehutch** [comment]
@cowwoc Interesting -- you're almost certainly right that `requires java.logging` will fix this, although like @GedMarc, I haven't seen this problem either when using ClassGraph as a module. I'll add it anyway though, since it seems like that's the right fix.

Unfortunately I'm in the middle of a mammoth refactoring effort right now, and I'm trying to eliminate static analysis warnings from four different static code analyzers too (I decided to get the code as squeaky clean as possible, and it's no small feat to keep four static analyzers happy, when they start out with hundreds of mostly overly pedantic warnings). So I suggest cloning 4.6.32 and building your own ClassGraph jar for now with the fix you suggested, since it might take a few days for me to be able to push out the next release (which will be 4.8.0).

@GedMarc I haven't figured out yet when Java decides to enforce encapsulation and when it doesn't. Certainly if you add ClassGraph to the traditional classpath, the module descriptor is ignored, but also it seems like sometimes encapsulation is not enforced on the module path.

**Comment by lukehutch** [comment]
> (y) verbose logging

@GedMarc -- ??

**Comment by lukehutch** [comment]
BTW I have a sneaking suspicion that having to add missing "requires" statements to module descriptors is going to quickly become an annoying part of a Java programmer's existence :-)

**Comment by GedMarc** [comment]
Double checked -- Need a test case, I can't replicate this, I don't require the exposure and no errors are thrown

JPMS, Strictly Named, JDK 11, Module-Info


[test_verbose_requirements.zip](https://github.com/classgraph/classgraph/files/2848566/test_verbose_requirements.zip)

**Comment by cowwoc** [comment]
Guys, this issue is somehow related to https://youtrack.jetbrains.com/issue/IDEA-171320.

I only get this error when using IDEA's built-in runner. Maven works fine. 

What's worse, I tried invoking "java --list-modules" and `java.logging` is included by default.

I'm going to try putting together a minimal testcase so we can proceed further.

**Comment by GedMarc** [comment]
I use IntelliJ - 

Let me send you my full test, maybe you can replicate the ide thing in here

P.S> Are you registering java.logging in your module-info?

[test_verbose_intellij.zip](https://github.com/classgraph/classgraph/files/2848571/test_verbose_intellij.zip)

**Comment by GedMarc** [comment]
ok, so java.logging is never included by default - don't know where you got that from -

The parent module must define the necessities of all underlying unless they are "requires transitive" - i think you are just using it wrong?

in YOUR module-info, add in requires java.logging so it is loaded into the parent module.

**Comment by cowwoc** [comment]
@GedMarc Sorry, you're right. `java.logging` is observable by default, but not automatically imported. Anyway, I tried adding `requires java.logging` to my own module and it did not help. And by your own reasoning: if `java.logging` is not auto-imported then ClassGraph needs to add this line to its `module-info.java`.

**Comment by GedMarc** [comment]
What does the test project do in your intellij?

You are partially right, But if the parent doesn't declare that it is used, then you will get an error in your module-info saying it is not specified (I've done this dance so many times now) - this is why transient exists

I would say @lukehutch - make it requires static java.logging - it is only needed when verbose is on,
@cowwoc No matter what you do, you will need to register requires java.logging it in your parent module

P.S> In JDK 12 it is System.logging

**Comment by lukehutch** [comment]
> `java.logging` is "observable" by default, but not automatically imported.

@cowwoc What does this mean? The Java definition of observable module is:

> The modules built-in to the compile-time or run-time environment, together with those defined by artifacts on module paths, are collectively referred to as the universe of *observable modules*.

**Comment by lukehutch** [comment]
> I would say make it requires static java.logging - it is only needed when verbose is on

This is reasonable, but given that it is a `java.util` class, won't it always be present in any JRE?

**Comment by GedMarc** [comment]
@lukehutch One would think hey xD

The reasoning is it is java.util.logging - so not directly part of java.base/java.utilities xD
It's packaged as java.logging for now, until it becomes system.logging -

The idea from what i've read on this, is that all custom loggers (sl4j, jboss-logging etc etc etc) shouldn't be used (what a nasty way of doing it!), as the jdk9 update to the logging library caters for "everything"

Haven't come across a scenario where it doesn't fit yet, so I guess they are right

**Comment by cowwoc** [comment]
> What does this mean? The Java definition of observable module is:

I just meant that `java.logging` is available for importing (`requires`) but is not imported by default.

**Comment by lukehutch** [comment]
See the quote I just added above (might have crossed mid-air):

> As a consequence even if a module is present on the module path (or in the JDK for that matter), it will not be added to the module graph just because of an optional dependency. It will only make it into the graph if it is also a regular dependency of some other module that is being resolved or because it was added explicitly with the command line flag --add-modules.

**Comment by cowwoc** [comment]
@GedMarc P.S> In JDK 12 it is System.logging

Where do you see this? If I invoke `java --list-modules` under JDK 12-ea+28 I still see it under `java.logging`.

**Comment by GedMarc** [comment]
@cowwoc https://docs.oracle.com/javase/9/docs/api/java/lang/System.Logger.html -- This guy

He's been trying to get into the jdk mainstream for ages now, JDK 12 he was earmarked for.. Let me check the JEPs

**Comment by GedMarc** [comment]
@lukehutch I don't think you should remove it, it certainly is (i think) the correct approach for optional logging -

 It will only make it into the graph if it is also a regular dependency of some other module
-- They are saying here you will still have to mark it as requires in YOUR module-info to use it, but the library will work fine without it

 or because it was added explicitly with the command line flag --add-modules.
Command line addition into the parent module (if the module it is calling from doesn't specify it explicitely)

**Comment by lukehutch** [comment]
> P.S In JDK 12 it is System.logging

@cowwoc Ouch, are they removing java.logging? That would break everything under the sun. Or are those two going to overlap for a while?

@GedMarc It looks like the right way to use `requires static xyz` is to use it if and only if `xyz` is only required by your library when code that is calling your library `requires xyz`. This is not strictly the case for `.verbose()` in ClassGraph.

**Comment by cowwoc** [comment]
Here is my testcase: [classgraph.zip](https://github.com/classgraph/classgraph/files/2848578/classgraph.zip)

Please let me know if you are able to reproduce the problem with it.

**Comment by GedMarc** [comment]
reproduced -

**Comment by GedMarc** [comment]
OK - it works in my test case because junit.jupitor has a requires transient java.logging -

I'd say please use requires static java.logging so that it isn't a mandatory rquirement (verbose would never happen in produciton)

**Comment by lukehutch** [comment]
I still think I should remove `static`, and the reason is that even just writing the log to stderr now goes through `java.logging`, even if a logging framework is not present. I don't want to have to wrap that in reflection with fallback to `System.err.println()`, which is the only way I would be comfortable adding a static `requires`.

**Comment by GedMarc** [comment]
Cool :) 👍

**Comment by lukehutch** [comment]
Remember that a goal of ClassGraph is to have no external dependencies. I'm fine liberally adding JRE dependencies as a consequence of the ton of work I did to rewrite / reinvent every wheel to avoid adding even a single non-JRE dependency!

**Comment by lukehutch** [comment]
@cowwoc I'm still interested in the `java.logging` vs. `system.logging` thing, since I hope that ClassGraph ins't going to break at some time soon and have to support two logging systems. However my guess is that for a long time, `java.logging` will proxy over to `system.logging`, and that new code will be written for `system.logging`, the way that `sun.misc.Unsafe` proxies over to `jdk.internal.misc.Unsafe`, and new code should use `jdk.internal.misc.Unsafe`. The JDK team care way too much about backwards compatibility not to do that.

However, this raises the question about whether I should require `system.logging` too. I think I'll add that just to be safe, so I don't have to fix that sometime in the near future.

**Comment by cowwoc** [comment]
@lukehutch GedMarc was the one who brought up `system.logging`. I know nothing about it :)

**Comment by lukehutch** [comment]
Oh, oops, you're right :-D  @GedMarc ...

**Comment by GedMarc** [comment]
@lukehutch https://www.baeldung.com/java-9-logging-api

Also - https://github.com/AdoptOpenJDK/openjdk-jdk11/tree/master/test/jdk/java/lang/System/Logger

OK, So I did a mistake make - java.util.logging is an extension of System.Logger.

System.out and System.err use System.Logger (not java.util.logging)

**Comment by cowwoc** [comment]
@GedMarc Sounds to me like the module name is `java.logging` and `System.Logger` is just a collection of packages, classes and interfaces. Hence, `requires java.logging` is the right way to go and will not be changing in JDK 12.

**Comment by lukehutch** [comment]
@cowwoc ah, makes sense that it's always fine to require interfaces. Thanks for confirming.

I'll close this now, since the change is in master, but feel free to keep discussing here if you want. Thanks for the report, @cowwoc!

**Comment by cowwoc** [comment]
@lukehutch Minor nitpick: https://github.com/classgraph/classgraph/commit/e1e0126db90fa5675d2d2f11daf87fa3b32aee8e#r32258749

**Comment by lukehutch** [comment]
@cowwoc Good catch, thanks!

**Comment by cowwoc** [comment]
I just verified the fix in `master` fixes the error in my application. Thank you!

**Comment by lukehutch** [comment]
Thanks for verifying!

**Comment by lukehutch** [comment]
Released in 4.8.0.

https://github.com/classgraph/classgraph/releases/tag/classgraph-4.8.0

Thanks for the bug report!
